urlaf656038a0_Salihan04_MOBLIMA_tgz-pJ8-com_moblima_util_dataTestJ8/src/com/moblima/businesslogic/SalesByDayBL.java

```diff
@@ -0,0 +1,118 @@
+package com.moblima.businesslogic;
+
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Hashtable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.moblima.dataaccess.BookingDAO;
+import com.moblima.dataaccess.BookingDAOImpl;
+import com.moblima.dataaccess.CineplexDAO;
+import com.moblima.dataaccess.CineplexDAOImpl;
+import com.moblima.dataaccess.ShowTimeDAO;
+import com.moblima.dataaccess.ShowTimeDAOImpl;
+import com.moblima.model.Booking;
+import com.moblima.model.Cinema;
+import com.moblima.model.Cineplex;
+import com.moblima.model.Movie;
+import com.moblima.model.MovieTicket;
+import com.moblima.model.ShowTime;
+import com.moblima.model.Transaction;
+
+public class SalesByDayBL {
+
+	private Calendar calendar;
+	private CineplexDAO cineplexDAO;
+	private ShowTimeDAO showTimeDAO;
+	private List<Date> daysOfCurrentMonth;
+	private List<Cineplex> cineplexes;
+	private List<Movie> moviesByCineplex;
+	private int currentMonth;
+	private int showTimeMonth;
+	private float indiTakings;
+
+	public SalesByDayBL() {
+		calendar = Calendar.getInstance();
+		showTimeDAO = ShowTimeDAOImpl.getInstance();
+		cineplexDAO = CineplexDAOImpl.getInstance();
+		daysOfCurrentMonth = new ArrayList<Date>();
+		cineplexes = new ArrayList<Cineplex>();
+		moviesByCineplex = new ArrayList<Movie>();
+		currentMonth = calendar.get(Calendar.MONTH);
+		showTimeMonth = 0;
+		indiTakings = 0;
+	}
+
+	public List<Date> getDaysOfCurrentMonth() {
+
+		for(ShowTime showTime : showTimeDAO.getShowTimes()) {
+			Date date = showTime.getTime();
+			calendar.setTime(date);
+			showTimeMonth = calendar.get(Calendar.MONTH);
+			if(showTimeMonth == currentMonth && !daysOfCurrentMonth.contains(date))
+				daysOfCurrentMonth.add(date);
+		}
+
+		return daysOfCurrentMonth;
+	}
+
+	public List<Cineplex> getCineplexes(Date date) {
+		cineplexes = new ArrayList<Cineplex>();
+		for(Cineplex cineplex : cineplexDAO.getCineplexes()) {
+			for(Cinema c : cineplex.getCinemas()) {
+				for(ShowTime st : c.getShowTimes()) {
+					if(daysOfCurrentMonth.contains(st.getTime())) {
+						cineplexes.add(cineplex);
+						break;
+					}
+				}
+			}
+		}
+		return cineplexes;
+	}
+
+	public List<Movie> getMoviesByCineplex(Cineplex cineplex) {
+		moviesByCineplex = new ArrayList<Movie>();
+
+		for(Cinema c : cineplex.getCinemas()) {
+			for(ShowTime st : c.getShowTimes()) {
+				int month;
+				calendar.setTime(st.getTime());
+				month = calendar.get(Calendar.MONTH);
+				if(st.getMovieTickets().size() > 0 && !moviesByCineplex.contains(st.getMovie()) && currentMonth == month) {
+					moviesByCineplex.add(st.getMovie());
+				}
+			}
+		}
+
+		return moviesByCineplex;
+	}
+
+	public float getindiTakings(Date date, Movie movie) {
+
+		indiTakings = 0;
+
+		for(ShowTime st : movie.getShowTimes()) {
+			calendar.setTime(date);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;date&#39; (index: 0 | type: Date). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;movie&#39;. The same parameter type is validated here for example: urlaf656038a0_Salihan04_MOBLIMA_tgz-pJ8-com_moblima_util_dataTestJ8/src/com/moblima/businesslogic/ShowTimeBL.java:50. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).